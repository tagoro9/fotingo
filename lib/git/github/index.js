'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _child_process = require('child_process');

var _child_process2 = _interopRequireDefault(_child_process);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _github = require('github');

var _github2 = _interopRequireDefault(_github);

var _ramda = require('ramda');

var _ramda2 = _interopRequireDefault(_ramda);

var _util = require('../util');

var _util2 = require('../../util');

var _error = require('../../error');

var _reporter = require('../../reporter');

var _reporter2 = _interopRequireDefault(_reporter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var github = new _github2['default']({});

var getCurrentUser = _ramda2['default'].compose(_ramda2['default'].partial(_ramda2['default'].__, [{}]), _util2.promisify)(github.users.get);

var createPullRequest = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.pullRequests.create));
var addLabels = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.issues.addLabels));
var getLabels = _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].prop('data')), (0, _util2.promisify)(github.issues.getLabels));

var authenticate = _ramda2['default'].compose(_util2.wrapInPromise, function (token) {
  return github.authenticate(token);
}, _ramda2['default'].set(_ramda2['default'].lensProp('token'), _ramda2['default'].__, { type: 'token' }));

var readUserToken = _ramda2['default'].compose(_ramda2['default'].partial(_reporter2['default'].question, [{ question: 'Introduce a Github personal token' }]), (0, _util2.debugCurried)('github', 'Github account not set'));

var authenticateAndGetCurrentUser = _ramda2['default'].composeP(_ramda2['default'].compose((0, _error.catchPromiseAndThrow)('github', function (e) {
  (0, _util2.debug)('github', 'Error when authenticating: ' + e.message);
  switch (e.code) {
    case '500':
      return _error.errors.github.cantConnect;
    default:
      return _error.errors.github.tokenInvalid;
  }
}), getCurrentUser), authenticate);

// Object -> Object -> String
var buildPullRequestTitle = _ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].compose(_ramda2['default'].concat(_ramda2['default'].__, '\n'), _ramda2['default'].prop('name'), _ramda2['default'].nthArg(1)), function (issue) {
  return _util.ISSUE_TYPES[issue.fields.issuetype.name] + '/' + issue.key + ' ' + _ramda2['default'].take(60, issue.fields.summary) + '\n';
});

// Array -> String
var buildPullRequestBody = _ramda2['default'].compose(_ramda2['default'].join('\n'), _ramda2['default'].map(_ramda2['default'].converge(_ramda2['default'].concat, [_ramda2['default'].compose(_ramda2['default'].concat('* '), _ramda2['default'].prop('header')), _ramda2['default'].compose(_ramda2['default'].ifElse(_ramda2['default'].isNil, _ramda2['default'].always(''), _ramda2['default'].replace(/\n|^/g, '\n  ')), _ramda2['default'].prop('body'))])));
// Array -> String
var buildPullRequestFooter = function () {
  function buildPullRequestFooter(issueRoot) {
    return _ramda2['default'].compose(_ramda2['default'].ifElse(_ramda2['default'].isEmpty, _ramda2['default'].always(''), _ramda2['default'].compose(_ramda2['default'].concat('\nFixes '), _ramda2['default'].join(', '))), _ramda2['default'].map(function (_ref) {
      var raw = _ref.raw,
          issue = _ref.issue;
      return '[' + raw + '](' + issueRoot + issue + ')';
    }), _ramda2['default'].uniqBy(_ramda2['default'].prop('issue')));
  }

  return buildPullRequestFooter;
}();

// Object -> Object -> String
var buildPullRequestDescription = function () {
  function buildPullRequestDescription(issueRoot) {
    return _ramda2['default'].converge(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].unapply(_ramda2['default'].join('\n'))), [buildPullRequestTitle, _ramda2['default'].compose(buildPullRequestBody, _ramda2['default'].prop('commits'), _ramda2['default'].nthArg(1)), _ramda2['default'].compose(buildPullRequestFooter(issueRoot), _ramda2['default'].prop('issues'), _ramda2['default'].nthArg(1))]);
  }

  return buildPullRequestDescription;
}();

// String -> String -> *
var writeFile = _ramda2['default'].curryN(2, function (file, content) {
  return (0, _util2.promisify)(_fs2['default'].writeFile)(file, content, 'utf8');
});
// String -> * -> String
var readFile = function () {
  function readFile(file) {
    return function () {
      return (0, _util2.promisify)(_fs2['default'].readFile)(file, 'utf8');
    };
  }

  return readFile;
}();
// String -> String -> String
var deleteFile = function () {
  function deleteFile(file) {
    return function (content) {
      return (0, _util2.promisify)(_fs2['default'].unlink)(file).then(function () {
        return (0, _util2.wrapInPromise)(content);
      });
    };
  }

  return deleteFile;
}();
// * -> String
var editFile = function () {
  function editFile(file) {
    return function () {
      return new Promise(function (resolve, reject) {
        var vim = _child_process2['default'].spawn('vim', [file], { stdio: 'inherit' });
        vim.on('exit', _ramda2['default'].ifElse(_ramda2['default'].equals(0), resolve, reject));
      });
    };
  }

  return editFile;
}();

// Object -> String -> String
var allowUserToEditPullRequest = function () {
  function allowUserToEditPullRequest(description) {
    var prFile = '/tmp/fotingo-pr-' + Date.now();
    return _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].trim), deleteFile(prFile), readFile(prFile), editFile(prFile), writeFile(prFile))(description);
  }

  return allowUserToEditPullRequest;
}();

var submitPullRequest = _ramda2['default'].curryN(4, function (config, project, branchInfo, description) {
  return createPullRequest({
    owner: config.get(['github', 'owner']),
    repo: project,
    head: branchInfo.name,
    base: config.get(['github', 'base']),
    title: _ramda2['default'].compose(_ramda2['default'].head, _ramda2['default'].split('\n'))(description),
    body: _ramda2['default'].compose(_ramda2['default'].join('\n'), _ramda2['default'].tail, _ramda2['default'].split('\n'))(description)
  });
});

var addLabelsToPullRequest = _ramda2['default'].curryN(4, function (config, project, branchInfo, pullRequest) {
  return _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].always(pullRequest)), addLabels)({
    owner: config.get(['github', 'owner']),
    repo: project,
    number: pullRequest.number,
    labels: branchInfo.labels
  });
});

exports['default'] = {
  init: function () {
    function init(config) {
      return function () {
        (0, _util2.debug)('github', 'Initializing Github api');

        var doLogin = _ramda2['default'].composeP(authenticateAndGetCurrentUser, config.update(['github', 'token']), readUserToken);

        var configPromise = _ramda2['default'].isNil(config.get(['github', 'owner'])) ? _ramda2['default'].composeP(config.update(['github', 'owner']), _reporter2['default'].question)({ question: 'What\'s the github repository owner?' }) : (0, _util2.wrapInPromise)(config.get(['github', 'owner']));

        return configPromise.then(function () {
          if (config.isGithubLoggedIn()) {
            (0, _util2.debug)('github', 'User token is present. Using current authentication');
            return authenticateAndGetCurrentUser(config.get(['github', 'token']))
            // TODO differentiate error codes so only login is attempted when tokenInvalid
            ['catch'](_ramda2['default'].composeP(doLogin, (0, _util2.debugCurriedP)('github', 'Current authentication failed. Attempting login')));
          }
          (0, _util2.debug)('github', 'No user token present. Attempting login');
          return doLogin();
        });
      };
    }

    return init;
  }(),
  // Object -> Array -> Promise
  createPullRequest: _ramda2['default'].curryN(5, function (config, project, issue, issueRoot, branchInfo) {
    return _ramda2['default'].composeP(_ramda2['default'].ifElse(_ramda2['default'].isEmpty, (0, _error.throwControlledError)(_error.errors.github.pullRequestDescriptionInvalid),
    // Assign the PR link to all the issues that were created
    _ramda2['default'].composeP(_ramda2['default'].compose(_util2.wrapInPromise, _ramda2['default'].set(_ramda2['default'].lensProp('pullRequest'), _ramda2['default'].__, { branchInfo: branchInfo })), addLabelsToPullRequest(config, project, branchInfo), submitPullRequest(config, project, branchInfo))), (0, _util2.debugCurriedP)('github', 'Submitting pull request to github'), allowUserToEditPullRequest, (0, _util2.debugCurriedP)('github', 'Editing pull request description'), buildPullRequestDescription(issueRoot))(issue, branchInfo);
  }),
  checkAndGetLabels: _ramda2['default'].curryN(4, function (config, project, labels, branchInfo) {
    return _ramda2['default'].composeP(_ramda2['default'].compose(_ramda2['default'].set(_ramda2['default'].lensProp('labels'), _ramda2['default'].__, branchInfo), _ramda2['default'].map(_ramda2['default'].prop('name')), _ramda2['default'].filter(_ramda2['default'].compose(_ramda2['default'].contains(_ramda2['default'].__, labels), _ramda2['default'].prop('name')))), _ramda2['default'].unless(_ramda2['default'].compose(function (names) {
      return _ramda2['default'].all(_ramda2['default'].contains(_ramda2['default'].__, names), labels);
    }, _ramda2['default'].map(_ramda2['default'].prop('name'))), (0, _error.throwControlledError)(_error.errors.github.invalidLabel)), getLabels)({ owner: config.get(['github', 'owner']), repo: project });
  })
};