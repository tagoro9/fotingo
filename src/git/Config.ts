/**
 * Git config
 */
export interface GitConfig {
  /**
   * Name of the base branch to use when creating a new branch
   */
  baseBranch: string;
  /**
   * Template to use for creating branch names
   */
  branchTemplate: string;

  /**
   * Name of the remote to connect to
   */
  remote: string;
}

export interface GithubConfig {
  authToken: string;
  baseBranch: string;
  owner: string;
  pullRequestTemplate: string;
  repo: string;
}
