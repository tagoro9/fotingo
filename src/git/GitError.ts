export enum GitErrorType {
  BRANCH_ALREADY_EXISTS,
  NOT_A_GIT_REPO,
}

export interface GitError {
  readonly code: GitErrorType;
  readonly message: string;
}

export class GitErrorImpl extends Error implements GitError {
  public readonly code: GitErrorType;

  constructor(message: string, code: GitErrorType) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.code = code;
  }
}
