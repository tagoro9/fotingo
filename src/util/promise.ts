/**
 * Given a list of promise providers, execute them in series and return a promise that resolves with
 * all the resolved values from the providers
 * @param providers Promise providers
 */
export function series<T>(providers: Array<() => Promise<T>>): Promise<T[]> {
  const ret = Promise.resolve(null);
  const results: T[] = [];

  return providers
    .reduce((result, provider, index) => {
      return result.then(() => {
        return provider().then(val => {
          results[index] = val;
        });
      });
    }, ret)
    .then(() => {
      return results;
    });
}
