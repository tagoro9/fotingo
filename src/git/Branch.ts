/**
 * Git branch handler
 */
import * as escapeStringRegexp from 'escape-string-regexp';
import {
  __,
  compose,
  converge,
  curryN,
  identity,
  ifElse,
  isNil,
  lt,
  mapObjIndexed,
  match,
  nth,
  nthArg,
  prop,
  reduce,
  replace,
  subtract,
  toPairs,
  toUpper,
} from 'ramda';
import { Issue, IssueType } from 'src/types';
import { GitConfig } from './Config';

const ISSUE_TYPE_TO_BRANCH_PREFIX: { [S in IssueType]: string } = {
  [IssueType.BUG]: 'b',
  [IssueType.FEATURE]: 'f',
  [IssueType.STORY]: 'f',
  [IssueType.SUB_TASK]: 'f',
  [IssueType.TASK]: 'c',
};

enum TemplateKey {
  ISSUE_SHORT_NAME = 'issue.shortName',
  ISSUE_KEY = 'issue.key',
  ISSUE_SANITIZED_SUMMARY = 'issue.sanitizedSummary',
}

const TEMPLATE_KEYS_TO_MATCHERS: { [S in TemplateKey]: string } = {
  [TemplateKey.ISSUE_SHORT_NAME]: '(\\w+)',
  [TemplateKey.ISSUE_KEY]: '(\\w+-\\d+)',
  [TemplateKey.ISSUE_SANITIZED_SUMMARY]: '([\\w_-]+)',
};

interface TemplateData {
  data: {
    [TemplateKey.ISSUE_SHORT_NAME]: string;
    [TemplateKey.ISSUE_KEY]: string;
    [TemplateKey.ISSUE_SANITIZED_SUMMARY]: string;
  };
  template: string;
}

/**
 * Given a git configuration and an issue build the branch template data
 * @param config Git configuration
 * @param issue Issue
 */
function getTemplateData(config: GitConfig, issue: Issue): TemplateData {
  return {
    data: {
      [TemplateKey.ISSUE_SHORT_NAME]: ISSUE_TYPE_TO_BRANCH_PREFIX[issue.type],
      [TemplateKey.ISSUE_KEY]: issue.key.toLowerCase(),
      [TemplateKey.ISSUE_SANITIZED_SUMMARY]: issue.sanitizedSummary,
    },
    template: config.branchTemplate,
  };
}

/**
 * Get a branch name for the passed configuration and issue
 * @param config Git configuration
 * @param issue Issue
 */
export const getName = curryN(
  2,
  compose(
    converge(
      reduce((msg: string, [k, v]: string[]) => replace(`{${k}}`, v, msg)),
      [prop('template'), compose(toPairs, prop('data'))],
    ),
    getTemplateData,
  ),
);

/**
 * Build a branch template regex from the current git config branch template
 * @param config Git configuration
 */
const buildBranchTemplateRegex = compose(
  reduce(
    (msg: string, [k, v]: [TemplateKey, string]) => replace(`{${k}}`, `${v}`, msg),
    __ as any,
    toPairs(TEMPLATE_KEYS_TO_MATCHERS),
  ),
  prop('branchTemplate'),
);

/**
 * Given a regex that matches a branchTemplate, build a map between the template keys and the
 * matching group index in the regex
 * @param regex Regex to match a branch name
 */
const getTemplateKeysMatchIndexMap = (regex: string): { [S in TemplateKey]: number } => {
  const indexMap = Object.values(TemplateKey).reduce(
    (acc, val: TemplateKey) => ({
      ...acc,
      [val]: (
        regex.match(escapeStringRegexp(TEMPLATE_KEYS_TO_MATCHERS[val])) || {
          index: -1,
        }
      ).index,
    }),
    {},
  );
  const orderedValues = Object.values(indexMap)
    .sort(subtract)
    .filter(lt(-1));
  return mapObjIndexed(i => 1 + orderedValues.indexOf(i), indexMap) as {
    [S in TemplateKey]: number;
  };
};

/**
 * Extract the issue id from the branch name given a git configuration
 * @param config Git configuration
 * @param branchName Branch name
 */
export const getIssueId = curryN(
  2,
  compose(
    converge(
      compose(
        ifElse(isNil, identity, toUpper),
        converge(nth, [compose(prop(TemplateKey.ISSUE_KEY), getTemplateKeysMatchIndexMap), match]),
      ),
      [buildBranchTemplateRegex, nthArg(1)],
    ),
  ),
);
