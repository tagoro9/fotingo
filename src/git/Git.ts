import { boundMethod } from 'autobind-decorator';
import { CommitReference, ParsedCommit, sync } from 'conventional-commits-parser';
import createDebugger from 'debug';
import * as gitUrlParse from 'git-url-parse';
import {
  always,
  compose,
  concat,
  converge,
  filter,
  flatten,
  ifElse,
  isNil,
  join,
  map,
  nthArg,
  prop,
  propEq,
  props,
  replace,
  trim,
  uniqBy,
} from 'ramda';
import * as simpleGit from 'simple-git/promise';
import { BranchSummary } from 'simple-git/typings/response';
import { maybeAskUserToSelectMatches } from 'src/io/input-util';
import { Emoji, Messenger } from 'src/io/messenger';
import { findMatches } from 'src/io/text-util';
import { Issue } from 'src/issue-tracker/Issue';

import { getIssueId, getName } from './Branch';
import { GitConfig } from './Config';
import { GitErrorImpl, GitErrorType } from './GitError';
import { GitRemote } from './Remote';

const debug = createDebugger('fotingo:git');

interface Remote {
  name: string;
  refs: {
    fetch: string;
    push: string;
  };
}

interface GitLogLine {
  hash: string;
  date: string;
  message: string;
  author_name: string;
  author_email: string;
}
interface GitLog {
  latest: GitLogLine;
  all: GitLogLine[];
  total: number;
}

interface GitStatus {
  files: string[];
}

export interface BranchInfo {
  name: string;
  commits: ParsedCommit[];
  issues: CommitIssue[];
}

interface CommitIssue {
  raw: 'string';
  issue: string;
}

export class Git {
  private git: any;
  private config: GitConfig;
  private messenger: Messenger;

  constructor(config: GitConfig, messenger?: Messenger) {
    this.git = simpleGit().silent(true);
    this.config = config;
    // TODO This is error prone
    if (messenger) {
      this.messenger = messenger;
    }
  }

  /**
   * Given an issue, get the associated branch name
   * @param issue Issue
   */
  @boundMethod
  public getBranchNameForIssue(issue: Issue): string {
    debug(`creating branch name for ${issue.key}`);
    return getName(this.config, issue);
  }

  /**
   * Create a branch for the given name stashing any pending change
   * in the repo
   * @param branchName Branch na,e
   */
  @boundMethod
  public createBranchAndStashChanges(branchName: string): Promise<any> {
    return this.fetch()
      .then(() => this.doesBranchExist(branchName))
      .then(exists => {
        if (exists) {
          // TODO Imrpove this error
          throw new Error('There is already a branch for the issue');
        }
      })
      .then(() => this.maybeStashChanges())
      .then(() => this.findBaseBranch())
      .then(baseBranch => this.getLatestCommit(baseBranch).then(log => log.latest.hash))
      .then(lastCommitHash => this.git.checkoutBranch(branchName, lastCommitHash))
      .catch(this.mapAndThrowError);
  }

  /**
   * Push the current branch. If it doesn't exist on the remote, then publish it
   */
  @boundMethod
  public async push(): Promise<void> {
    const remoteExists = await this.doesCurrentBranchExistInRemote();
    if (remoteExists) {
      this.messenger.emit('Pushing branch', Emoji.ARROW_UP);
      this.messenger.inThread(true);
      await this.git.push(this.config.remote);
      this.messenger.inThread(false);
      return;
    }
    this.messenger.emit('Publishing branch', Emoji.ARROW_UP);
    this.messenger.inThread(true);
    await this.publish();
    this.messenger.inThread(false);
    return;
  }

  /**
   * Get full information about the current branch (including commits and fixed issues)
   */
  @boundMethod
  public async getBranchInfo(): Promise<BranchInfo> {
    const branchName = await this.getCurrentBranchName();
    const issueId = getIssueId(this.config, branchName);
    this.messenger.emit('Analyzing commit history', Emoji.MAG_RIGHT);
    const commits = await this.getBranchCommitsFromMergeBase().then(this.transformCommits);
    return {
      commits,
      issues: this.getIssues(commits, issueId),
      name: branchName,
    };
  }

  /**
   * Get the remote information for the given remote name
   * @param name Remote name
   */
  public getRemote(name: string): Promise<GitRemote> {
    return this.git.getRemotes(true).then((remotes: Remote[]) => {
      const origin = remotes.find((remote: Remote) => remote.name === name);
      const firstRemote = remotes[0];

      if (!origin && !firstRemote) {
        return Promise.reject();
      }

      return gitUrlParse((origin || firstRemote).refs.fetch);
    });
  }

  /**
   * Get the root dir of the repository
   */
  public getRootDir(): Promise<string> {
    return this.git
      .raw(['rev-parse', '--show-toplevel'])
      .then(compose(trim, replace('\n', '')))
      .catch(this.mapAndThrowError);
  }

  public async doesBranchExist(branchName: string): Promise<boolean> {
    return this.git
      .branchLocal()
      .then((branches: BranchSummary) => branches.all.some(branch => branch === branchName));
  }

  /**
   * Check if the current branch has a tracking branch in the configured remote
   */
  public async doesCurrentBranchExistInRemote(): Promise<boolean> {
    return this.git
      .revparse(['--abbrev-ref', '--symbolic-full-name', '@{u}'])
      .then(() => true)
      .catch((e: Error) => {
        if (/no upstream configured for branch/.test(e.message)) {
          return false;
        }
        this.mapAndThrowError(e);
      });
  }

  /**
   * Find the base branch based on the remote config and baseBranch prefix.
   * If none can be found, then throw an error
   */
  @boundMethod
  // TODO This is going to get called several times. It should be memoized
  public async findBaseBranch(removePrefix: boolean = false): Promise<string> {
    const branchPrefix = `remotes/${this.config.remote}`;
    const branches: Array<{ name: string }> = ((await this.git.branch(['-a'])) as BranchSummary).all
      .filter(b => b.startsWith(branchPrefix))
      .map((branchName: string) => ({ name: branchName }));

    const matches = findMatches(
      {
        checkForExactMatchFirst: true,
        cleanData: item => item.replace(`${branchPrefix}/`, ''),
        data: branches,
        fields: ['name'],
      },
      [this.config.baseBranch],
    );

    if (matches.length === 0) {
      throw new Error(
        `Could not find a branch in ${this.config.remote} that matches ${this.config.baseBranch}. Make sure that the branch is published in the remote`,
      );
    }

    const baseBranch = await maybeAskUserToSelectMatches(
      {
        data: matches,
        getLabel: branch => branch.name.replace(`${branchPrefix}/`, ''),
        getQuestion: item =>
          `We couldn't find a unique match for the base branch "${item}", which one best matches?`,
        getValue: branch => branch.name,
        options: [this.config.baseBranch],
        useDefaults: false,
      },
      this.messenger,
    );

    const name = baseBranch[0].name;
    return removePrefix ? name.replace(`${branchPrefix}/`, '') : name;
  }

  private async publish(): Promise<any> {
    const branchName = await this.getCurrentBranchName();
    return this.git.push(['-u', this.config.remote, branchName]);
  }

  private getIssues(commits: ParsedCommit[], issueId?: string): CommitIssue[] {
    return converge(
      compose<CommitIssue[], CommitIssue[], CommitIssue[], CommitIssue[]>(
        uniqBy((issue: CommitIssue) => issue.issue),
        concat,
      ),
      [
        compose(
          ifElse(isNil, always([]), key => [{ raw: `#${key}`, issue: key }]),
          nthArg(0),
        ),
        compose(
          flatten,
          map(
            compose(
              map((ref: CommitReference) => ({
                issue: ref.issue,
                raw: `${ref.prefix}${ref.issue}`,
              })),
              filter(propEq('action', 'Fixes')),
              prop('references'),
            ),
          ),
          nthArg(1),
        ),
      ],
    )(issueId, commits);
  }

  @boundMethod
  private transformCommits(commits: GitLogLine[]): ParsedCommit[] {
    return map(compose(sync, compose(join('\n'), props(['message', 'body']))))(commits);
  }

  /**
   * Get the name for the current branch
   */
  private getCurrentBranchName(): Promise<string> {
    return this.git.revparse(['--abbrev-ref', 'HEAD']).catch(this.mapAndThrowError);
  }

  /**
   * Get the latest commit for the specified git reference
   * @param what Git reference
   */
  private getLatestCommit(what: string): Promise<GitLog> {
    return this.git.log(['-n1', what]);
  }

  /**
   * Fetch from the fotingo configured remote
   */
  private async fetch(): Promise<void> {
    this.messenger.inThread(true);
    await this.git.fetch(this.config.remote);
    this.messenger.inThread(false);
  }

  /**
   * Get the git status
   */
  private status(): Promise<GitStatus> {
    return this.git.status();
  }

  /**
   * Stash all the current changes (including untracked files). The stash name will
   * be auto generated by fotingo
   */
  private maybeStashChanges(): Promise<string> {
    return this.status().then(st => {
      if (st.files.length > 0) {
        return this.git.stash(['save', '--include-untracked', 'Auto generated by fotingo']);
      }
      return Promise.resolve();
    });
  }

  /**
   * Transform errors coming from simple-git into
   * known errors
   * @param e Error
   */
  private mapAndThrowError(e: Error): void {
    if (e.message.match(/A branch named .* already exists/)) {
      throw new GitErrorImpl(e.message, GitErrorType.BRANCH_ALREADY_EXISTS);
    }
    if (e.message.match(/not a git repository/)) {
      throw new GitErrorImpl(e.message, GitErrorType.NOT_A_GIT_REPO);
    }
    throw e;
  }

  /**
   * Get the list of commits between the current branch HEAD
   * and the fotingo configured remote
   */
  private async getBranchCommitsFromMergeBase(): Promise<GitLogLine[]> {
    const baseBranch = await this.findBaseBranch();
    const ref = await this.git.raw(['merge-base', 'HEAD', baseBranch]);
    return this.git
      .log({
        from: 'HEAD',
        to: compose(trim, replace('\n', ''))(ref),
      })
      .then((data: GitLog) => data.all);
  }
}
