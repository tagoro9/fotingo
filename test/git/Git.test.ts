jest.mock('simple-git/promise');
import 'jest';
import * as simpleGitMock from 'simple-git/promise';
import { Git } from 'src/git/Git';
import { GitErrorType } from 'src/git/GitError';
import data from 'test/lib/data';

const simpleGit = (simpleGitMock as any) as jest.Mock;

let git: Git;
const issue = data.createIssue();
const gitConfig = data.createGitConfig();
const branchName = issue.key;
const latestCommitHash = '547433c';
const remoteBranch = `${gitConfig.remote}/${gitConfig.baseBranch}`;
const remote = 'git@github.com:tagoro9/fotingo-rewrite.git';

const gitMocks = {
  checkoutBranch: jest.fn().mockResolvedValue(undefined),
  fetch: jest.fn().mockResolvedValue(undefined),
  getRemotes: jest.fn().mockResolvedValue([
    {
      name: 'origin',
      refs: {
        fetch: remote,
        push: remote,
      },
    },
  ]),
  log: jest.fn().mockResolvedValue({
    latest: {
      hash: latestCommitHash,
    },
  }),
  push: jest.fn().mockResolvedValue(undefined),
  revparse: jest.fn().mockResolvedValue(branchName),
  stash: jest.fn().mockResolvedValue(undefined),
  status: jest.fn().mockResolvedValue({
    files: [],
  }),
};

describe('Git', () => {
  beforeEach(() => {
    simpleGit.mockReturnValue(gitMocks);
    git = new Git(gitConfig);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('getBranchNameForIssue', () => {
    it('should generate a branch name for an issue', () => {
      expect(git.getBranchNameForIssue(issue)).toMatchSnapshot();
    });
  });

  describe('getIssues', () => {
    it('should get the issue from the branch name', async () => {
      const issueId = await git.getIssues();
      expect(issueId).toBe(issue.key);
      expect(gitMocks.revparse).toBeCalledWith(['--abrev-ref', 'HEAD']);
    });
  });

  describe('push', () => {
    it('should do a git push', async () => {
      await git.push();
      expect(gitMocks.push).toBeCalledTimes(1);
    });

    it('should throw an error if there is no upstream branch', async () => {
      gitMocks.push.mockRejectedValue(new Error('has no upstream branch'));
      try {
        await git.push();
      } catch (e) {
        expect(e.message).toBe('NO_UPSTREAM');
      }
    });
  });

  describe('createBranchAndStashChanges', () => {
    it('should fetch and create a branch from the latest commit', async () => {
      await git.createBranchAndStashChanges(branchName);
      expect(gitMocks.fetch).toHaveBeenCalledTimes(1);
      expect(gitMocks.status).toHaveBeenCalledTimes(1);
      expect(gitMocks.log).toHaveBeenCalledWith(['-n1', remoteBranch]);
      expect(gitMocks.checkoutBranch).toHaveBeenCalledWith(branchName, latestCommitHash);
      expect(gitMocks.stash).not.toHaveBeenCalled();
    });

    it('should stash any existing change before creating the branch', async () => {
      gitMocks.status.mockResolvedValue({ files: ['some file.txt'] });
      await git.createBranchAndStashChanges(branchName);
      expect(gitMocks.fetch).toHaveBeenCalledTimes(1);
      expect(gitMocks.stash).toHaveBeenCalledWith([
        'save',
        '--include-untracked',
        'Auto generated by fotingo',
      ]);
      expect(gitMocks.status).toHaveBeenCalledTimes(1);
      expect(gitMocks.log).toHaveBeenCalledWith(['-n1', remoteBranch]);
      expect(gitMocks.checkoutBranch).toHaveBeenCalledWith(branchName, latestCommitHash);
    });

    it('should throw an error if a branch already exists', async () => {
      gitMocks.checkoutBranch.mockRejectedValue(
        new Error(`A branch named ${branchName} already exists`),
      );
      try {
        await git.createBranchAndStashChanges(branchName);
      } catch (e) {
        expect(e.code).toBe(GitErrorType.BRANCH_ALREADY_EXISTS);
      }
    });
  });

  describe('getRemote', () => {
    it('should return the parsed remote', async () => {
      await expect(git.getRemote('origin')).resolves.toMatchSnapshot();
      expect(gitMocks.getRemotes).toHaveBeenCalledWith(true);
    });

    it('should fall back to the first remote if it cannot find the specified one', async () => {
      await expect(git.getRemote('some_remote')).resolves.toMatchSnapshot();
      expect(gitMocks.getRemotes).toHaveBeenCalledWith(true);
    });

    it('should throw an error if there are no remotes', async () => {
      gitMocks.getRemotes.mockResolvedValue([]);
      await expect(git.getRemote('origin')).rejects.toBeUndefined();
    });
  });
});
