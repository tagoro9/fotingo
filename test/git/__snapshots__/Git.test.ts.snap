// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Git getBranchNameForIssue should generate a branch name for an issue 1`] = `"fotingo-2484"`;

exports[`Git getRemote should fall back to the first remote if it cannot find the specified one 1`] = `
Object {
  "commit": undefined,
  "filepath": "",
  "filepathtype": "",
  "full_name": "tagoro9/fotingo-rewrite",
  "git_suffix": true,
  "hash": "",
  "href": "git@github.com:tagoro9/fotingo-rewrite.git",
  "name": "fotingo-rewrite",
  "organization": "tagoro9",
  "owner": "tagoro9",
  "pathname": "/tagoro9/fotingo-rewrite.git",
  "port": null,
  "protocol": "ssh",
  "protocols": Array [],
  "query": Object {},
  "ref": "",
  "resource": "github.com",
  "search": "",
  "source": "github.com",
  "toString": [Function],
  "token": "",
  "user": "git",
}
`;

exports[`Git getRemote should return the parsed remote 1`] = `
Object {
  "commit": undefined,
  "filepath": "",
  "filepathtype": "",
  "full_name": "tagoro9/fotingo-rewrite",
  "git_suffix": true,
  "hash": "",
  "href": "git@github.com:tagoro9/fotingo-rewrite.git",
  "name": "fotingo-rewrite",
  "organization": "tagoro9",
  "owner": "tagoro9",
  "pathname": "/tagoro9/fotingo-rewrite.git",
  "port": null,
  "protocol": "ssh",
  "protocols": Array [],
  "query": Object {},
  "ref": "",
  "resource": "github.com",
  "search": "",
  "source": "github.com",
  "toString": [Function],
  "token": "",
  "user": "git",
}
`;
